ApiData.ts

```diff
@@ -0,0 +1,266 @@
+/** Debounce utility function to limit function execution frequency */
+function debounce<F extends (...args: any) => any>(func: F, waitFor: number) {
+	let timeout: number;
+
+	return (...args: Parameters<F>): Promise<ReturnType<F>> => {
+		clearTimeout(timeout);
+
+		return new Promise((resolve) => {
+			timeout = setTimeout(() => {
+				resolve(func(...args));
+			}, waitFor);
+		});
+	};
+}
+
+/** Constants for grid calculation
+* GRID_RATIO represents the ratio of the grid's height to the window's height.
+*/
+const GRID_RATIO = 9 / 20;
+const ROW_HEIGHT = 16;
+
+/** manage data and settings on the grid */
+class ApiData {
+
+	pageSize: number;
+	currentPage: number = 1;
+	data: GridData[] = [];
+	totalItems: number = 0;
+	columnNames: ColumnName[] = [];
+	maxGridHeight: number = 0;
+	firstVal: number = 0;
+	lastVal: number = -1;
+	maxRange: number;
+
+	constructor(pageSize: number) {
+		this.pageSize = pageSize;
+		this.maxRange = 0;
+
+	}
+
+	/** Initialize method to set up the grid */
+	initialize(): Promise<void> {
+		this.adjustGridHeight();
+		return this.recordCount()
+			.then(() => this.fetchColumns())
+			.then(() => this.fetchRecords())
+			.then(() => this.setupControls());
+	}
+	/** Fetch total record count from the server */
+	recordCount(): Promise<void> {
+		return this.fetchData('http://localhost:2050/recordCount')
+			.then((response: any) => {
+				const totalItems = response;
+				this.totalItems = totalItems;
+			})
+			.catch(error => {
+				console.error('Failed to fetch record count:', error);
+				alert('Failed to fetch record count.');
+			});
```

Comment:
So, by catching the error here, and more specifically, not throwing an error here, the calling function has no idea something went wrong, and will try and continue executing the rest of your code. This is usually a good thing, as we don't want a front-end app to break just because one part of it is broken.

But when I look at `initialize`, it eventually calls `fetchRecords`, and `fetchRecords` definitely does need `totalItems` to have a value. I kinda feel that your program should not try and call `fetchRecords` if `recordCount` failed. And for that you are going to have to throw an error here.